[WIP] refactor: watch test cleanup

karlkfi opened the pull request:
#### What type of PR is this?
/kind cleanup
/kind flake

#### What this PR does / why we need it:

- Fix flaky watch tests that were not waiting for the storage watcher to be started by the server watch endpoint
- Use testify in watch tests to make the tests easier to read/debug
- Remove unnecessary test names from errors

#### Does this PR introduce a user-facing change?
```release-note
NONE
```

#### Dependencies
- https://github.com/kubernetes/kubernetes/pull/131505
- https://github.com/kubernetes/kubernetes/pull/131854
- https://github.com/kubernetes/kubernetes/pull/131656
- https://github.com/kubernetes/kubernetes/pull/131676
- https://github.com/kubernetes/kubernetes/pull/131680
- https://github.com/kubernetes/kubernetes/pull/131706

karlkfi commented:
/assign @wojtek-t

karlkfi committed:
test: validate that watchers close without error

- Add channel test helpers to k8s.io/client-go/util/testing
- Use the new channel test helpers in the client and server watch
  tests to validate that the result channels closes without error
  when the client stops the watcher.
- Use the new channel test helpers in the client watcher decoder
  tests to validate that encoded watch events can be decoded and
  that the decoder errors with EOF when stopped asynchronously.

These new tests uncovered existing errors (added TODOs):
1.  The watch client doesn't close the response body when it
    encounters a NegotiateError.
2.  The watch client sometimes sends a decode error on the result
    channel after the client watcher has been stopped.
3.  StreamWatcher.receive uses %v instead of %w when wrapping errors.

siyuanfoundation commented:
@karlkfi Thanks for the refactor. Is it possible to break this huge PR to smaller PRs or even commits?

karlkfi commented:
> @karlkfi Thanks for the refactor. Is it possible to break this huge PR to smaller PRs or even commits?

Sure. I'll see what i can do.

Might have to mostly just be separate commits tho, because making changes to the file causes it to be linted, which requires fixing a lot of different things in the same PR.

karlkfi commented:
Split into three commits for a first pass. This modifies more files tho, which might surface other hidden linter errors.

I can probably split the third commit down a bit more, pulling out the sub-tests, close helpers, and testify usage.

karlkfi committed:
test: remove unnecessary test names from errors

karlkfi committed:
test: Fix flaky watch tests

- Wait for storage watcher to be created by the watcher server
  before trying to send events to it in the tests.

karlkfi committed:
test: use testify in watch tests

- Use require and assert to make the tests easier to read and debug

yongruilin commented:
/triage accepted

karlkfi commented:
I split up the commits a bit more.

karlkfi commented:
Surprisingly, the request methods didn't cause additional linter failures by themselves, so i pulled those out to a seperate PR: https://github.com/kubernetes/kubernetes/pull/131656 You can approve that one first or this whole one. Whichever you prefer. If that one merges first, I'll rebase this one.

aojea commented on staging/src/k8s.io/apiserver/pkg/endpoints/watch_test.go:
```diff
@@ -25,17 +25,15 @@ import (
 	"time"
 
-	"github.com/google/go-cmp/cmp"
+	"github.com/stretchr/testify/assert"
```
https://github.com/kubernetes/kubernetes/issues/130346

aojea commented:
> * Use testify in watch tests to make the tests easier to read/debug

this is not true https://github.com/kubernetes/kubernetes/issues/130346 , tests are not easier to read and debug
